README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# FILEST <img src="man/figures/filest_logo.png" align="right"/>

<!-- badges: start -->

[![R-CMD-check](https://github.com/kridsadakorn/filest/workflows/R-CMD-check/badge.svg)](https://github.com/kridsadakorn/filest/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/FILEST)](https://CRAN.R-project.org/package=FILEST)
[![codecov](https://codecov.io/gh/kridsadakorn/filest/branch/master/graph/badge.svg?token=41WJVZOP49)](https://codecov.io/gh/kridsadakorn/filest)
[![License:
MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![Travis build
status](https://travis-ci.com/kridsadakorn/filest.svg?branch=master)](https://travis-ci.com/kridsadakorn/filest)
[![DOI](https://zenodo.org/badge/333290990.svg)](https://zenodo.org/badge/latestdoi/333290990)
[![FOSSA
Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fkridsadakorn%2Ffilest.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fkridsadakorn%2Ffilest?ref=badge_shield)
<!-- badges: end -->

`FILEST` (Fine-Level Structure Simulator) is a population genetic
simulator. The simulator is able to generate synthetic datasets for
single-nucleotide polymorphisms (SNP) for multiple populations. The
genetic distances among populations can be set according to the Fixation
Index (Fst). This tool is able to simulate outlying individuals and
missing SNPs can be specified. For Genome-wide association study (GWAS),
disease status can be set in desired level according risk ratio.

## Installation

You can install the released version of FILEST from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("FILEST")
```

Alternatively, you can install the dev version of FILEST from
[Github](https://github.com/kridsadakorn/filest) with

``` r
install.packages("remotes")
remotes::install_github("kridsadakorn/filest", dependencies = TRUE)
```

## Document

You can see the reference manual from:
<https://www.biostatgen.org/filest/>

## Example

This best way to understand the package is the run the demo function and
edit the parameters in the setting file:

``` r
library(FILEST)

output_dir <- demo.filest()
#> Creating a setting file ... /var/folders/k6/3_1zxcc90dgd495vv1432wf40000gn/T//Rtmpo0wJhk/example1.txt
#> Generating the simulated data  to  ... /var/folders/k6/3_1zxcc90dgd495vv1432wf40000gn/T//Rtmpo0wJhk
#> Start [S0] at 2022-04-27 02:31:39
#> Setting file is : /var/folders/k6/3_1zxcc90dgd495vv1432wf40000gn/T//Rtmpo0wJhk/example1.txt
#> The simulated files will be saved in this directory: /var/folders/k6/3_1zxcc90dgd495vv1432wf40000gn/T//Rtmpo0wJhk/example1
#> Creating data file setting #1 - rep #1
#> Done - 0.0299999713897705078 secs
#> Writing data files setting #1 - rep #1
#> Done - 0.149560928344726562 secs
#> Creating status file setting #1 - rep #1
#> Done - 0.15028691291809082 secs
#> Estimating Fst setting #1 - rep #1
#> Done - 0.163115024566650391 secs
#> Creating maker information setting #1 - rep #1
#> Done - 0.320117950439453125 secs
#> Generating PC scores #1 - rep #1
#> Generating EigenVector  #1 - rep #1
#> Done - 0.678966999053955078 secs
```

The demo function creates the setting file at a temp directory as
`example1.txt`, and you can use as a template to edit:

``` r
print(file.path(output_dir,"example1.txt"))
#> [1] "/var/folders/k6/3_1zxcc90dgd495vv1432wf40000gn/T//Rtmpo0wJhk/example1.txt"
```

The demo function create the simulated files at a temp directory as
\`well.`example1`.

``` r
dir(file.path(output_dir,"example1"))
#>  [1] "simSNP_rep1_data_numMark_rowInd_colVar.txt"
#>  [2] "simSNP_rep1_data_numMark_rowVar_colInd.txt"
#>  [3] "simSNP_rep1_eigenvector10.txt"             
#>  [4] "simSNP_rep1_estimated_Fst.txt"             
#>  [5] "simSNP_rep1_individuals_with_header.txt"   
#>  [6] "simSNP_rep1_individuals.txt"               
#>  [7] "simSNP_rep1_PC.pdf"                        
#>  [8] "simSNP_rep1_PC10.txt"                      
#>  [9] "simSNP_rep1.bed"                           
#> [10] "simSNP_rep1.bim"                           
#> [11] "simSNP_rep1.fam"                           
#> [12] "simSNP_rep1.RData"
```

## License

[![FOSSA
Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fkridsadakorn%2Ffilest.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fkridsadakorn%2Ffilest?ref=badge_large)

## About

-   Prof. Kristel Van Steen, visit
    <a href="http://bio3.giga.ulg.ac.be/" border=0 style="border:0; text-decoration:none; outline:none"><img src="man/figures/bio3_logo.png" width="40px" align="center"/></a><br />
-   Kridsadakorn Chaichoompu, visit
    <a href="https://www.biostatgen.org/" border=0 style="border:0; text-decoration:none; outline:none"><img src="man/figures/biostatgen_logo.png" width="110px" align="center"/></a><br />


NEWS.md

# FILEST 1.2.0

* Add the new way to random the value for outliers because the previous cannot 
use that the values will be different from the old values

# FILEST 1.1.3

* Update URL of Gitlab in description

# FILEST 1.1.2

* Update the package according to R v4

# FILEST 1.1.1

* Added the complete libraries in FILEST_package.R

# FILEST 1.1.0

* Added function create.setting.file


# FILEST 1.0.3

## Updates

* Dropped off all RD files for internal functions


# FILEST 1.0.2

## Updates

* Added references in the DESCRIPTION file
* Added single quotes for Software's names in the DESCRRIPTION file
* Added checking for output directory and input file


# FILEST 1.0.1

## Updates

* FILEST 1.0.1 is released under GPL-2|GPL-3

# FILEST 1.0.0

## NOTES

* Shift from 0.3.7 to 1.0.0 to release

## Updates

* Romoved all codes related to doMC and doParallel, there is a problem in Windows.

# FILEST 0.3.7

## Updates

* Added all manual files for all functions
* Added all unit tests

## Changes

* Removed save.PC.plot() and replaced with KRIS::plot3views

# FILEST 0.3.6

## Initial project in GitLab

* Uploaded and Synchronized all files


tests/testthat/test_filest.R
context("Test filest")

test_that("filest",{

  res <- demo.filest()

  expect_type(res, "character")
  expect_equal(file.exists(file.path(res,"example1.txt")), TRUE)
  expect_equal(dir.exists(file.path(res,"example1")), TRUE)
  expect_equal(file.exists(file.path(res,"example1","simSNP_rep1.bed")), TRUE)

  unlink(file.path(res,"example1"), recursive = TRUE)
  file.remove(file.path(res,"example1.txt"))

})

context("Test create.template.setting")

test_that("create.template.setting",{

  output <- file.path(tempdir(),"example_setting.txt")
  res <- create.template.setting(out.file = output, no.setting = 2)

  expect_type(res, "character")
  expect_equal(file.exists(res), TRUE)

  file.remove(res)

})


tests/testthat/test_internal_functions.R
context("Test cal.margin1 ")

test_that("Test cal.margin1",{

  res <- cal.margin1(0,1)
  expect_equal(res,0)

  res <- cal.margin1(1,1)
  expect_equal(res,0)

  res <- cal.margin1(1,2)
  expect_equal(res,-0.5)

  res <- cal.margin1(1,0.5)
  expect_equal(res,1)
})

context("Test cal.margin2")

test_that("Test cal.margin2",{

  res <- cal.margin2(0,1)
  expect_equal(res,0)

  res <- cal.margin1(1,1)
  expect_equal(res,0)

  res <- cal.margin2(1,2)
  expect_equal(res,0)

  res <- cal.margin2(1,0.5)
  expect_equal(res,0)

  res <- cal.margin2(0,0.5)
  expect_equal(res,1)

  res <- cal.margin2(0,2)
  expect_equal(res,-0.5)
})


context("Test get.random.beta")

test_that("Test get.random.beta",{

  res <- get.random.beta(0,1)
  expect_type(res,"double")

})


context("Test cal.prob.AA ")

test_that("Test cal.prob.AA",{

  res <- cal.prob.AA(0)
  expect_equal(res,1)

  res <- cal.prob.AA(1)
  expect_equal(res,0)

  res <- cal.prob.AA(0.5)
  expect_equal(res,0.25)

})


context("Test cal.prob.AB ")

test_that("Test cal.prob.AB",{

  res <- cal.prob.AB(0)
  expect_equal(res,0)

  res <- cal.prob.AB(1)
  expect_equal(res,0)

  res <- cal.prob.AB(0.5)
  expect_equal(res,0.5)

})


context("Test cal.prob.BB ")

test_that("Test cal.prob.BB",{

  res <- cal.prob.BB(0)
  expect_equal(res,0)

  res <- cal.prob.BB(1)
  expect_equal(res,1)

  res <- cal.prob.BB(0.5)
  expect_equal(res,0.25)

})


context("Test do.sample.SNP")

test_that("Test do.sample.SNP",{

  res <- do.sample.SNP(0,0.5,0.5,5)
  expect_length(res,5)

})


context("Test do.create.cate")

test_that("Test do.create.cate",{

  res <- do.create.cate("M","P")
  expect_equal(res,"P_M")

})


context("Test create.outlier")

test_that("Test create.outlier",{

  X <- matrix(c(1,2,0,1,2,2,1,2,0,0,1,2,1,2,2,2),ncol=4)
  P <- c(3,1)
  O <- c(0,1)

  E <- matrix(c(1,2,0,1,2,2,0,2,0,1,1,2,0,0,0,0),ncol=4, byrow = T)

  res <- create.outlier(X,P,O)
  expect_equal(res,E)

})



context("Test cal.prob.categorical.data")

test_that("Test cal.prob.categorical.data",{

  P <- 0.5
  M <- 2
  res <- cal.prob.categorical.data(P,M)
  expect_length(res,2)

  M <- 4
  res <- cal.prob.categorical.data(P,M)
  expect_length(res,4)

  M <- 6
  res <- cal.prob.categorical.data(P,M)
  expect_length(res,6)

})



context("Test do.sample.categorical.data")

test_that("Test do.sample.categorical.data",{

  P <- c(0.5,0.5,0.5,0.5,0.5)
  M <- 2
  I <- 1
  res <- cal.prob.categorical.data(P,M)
  expect_length(res,10)

  P <- c(0.5,0.5)
  M <- 2
  I <- 1
  res <- cal.prob.categorical.data(P,M)
  expect_length(res,4)

})



context("Test generate.categorical.data")

test_that("Test generate.categorical.data",{

  P <- 2
  Fst <- 0.1
  N <- 5
  res <- generate.categorical.data(P,Fst,N)
  expect_length(res,10)

})



context("Test generate.snp")

test_that("Test generate.snp",{

  P <- 2
  Fst <- 0.1
  N <- 5
  R <- 1.0
  res <- generate.snp(P,Fst,N,R)
  expect_length(res,10)

})



context("Test generate.label")

test_that("Test generate.label",{

  P <- c(5,5)
  O <- c(0,1)
  res <- generate.label(P,O)
  expect_length(res,10)

})



context("Test rbind_bigmatrix")

test_that("Test rbind_bigmatrix",{

  X <- matrix(c(1,2,0,1,2,2,1,2,0,0,1,2,1,2,2,2),ncol=4)
  Y <- matrix(c(1,2,0,1,2,2,1,2,0,0,1,2,1,2,2,2),ncol=4)
  res <- rbind_bigmatrix(X,Y)
  expect_length(res,32)

})



context("Test cbind_bigmatrix")

test_that("Test cbind_bigmatrix",{

  X <- matrix(c(1,2,0,1,2,2,1,2,0,0,1,2,1,2,2,2),ncol=4)
  Y <- matrix(c(1,2,0,1,2,2,1,2,0,0,1,2,1,2,2,2),ncol=4)
  res <- cbind_bigmatrix(X,Y)
  expect_length(res,32)

})


R/fst.R
##########################################################################
## IPCAPS Library
## Author: Kridsadakorn Chaichoompu (kridsadakorn.chaichoompu@ulg.ac.be)
## Date: 01/06/2016
## Description:
##    This code is a part of Iterative Pruning to CApture Population 
##    Structure (IPCAPS) Library
##    
##Licence: GPL V3
## 
##    Copyright (C) 2016  Kridsadakorn Chaichoompu
##
##    This program is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    This program is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with this program.  If not, see <http://www.gnu.org/licenses/>.

#--include--begin--#

#raw.data is a SNP matrix while rows represent individuals and columns represent SNP
#idx.p1 is a vector containing the indexes of population 1 in raw.data
#idx.p2 is a vector containing the indexes of population 2 in raw.data
#return a global Fst as a single value
fst.hudson <-function(raw.data,idx.p1,idx.p2){
  prestep.fst.one.marker <- function(alleles,idx.p1,idx.p2){
    
    #Pop 1
    G = alleles[idx.p1]
    no.AA=length(which(G==0))
    no.AB=length(which(G==1))
    no.BB=length(which(G==2))
    n1=(no.AA+no.AB+no.BB)*2
    p.A=(no.AA*2 + no.AB)/n1
    #p.B=(no.BB*2 + no.AB)/n1
    #p1 = min(p.A,p.B,na.rm = T)
    p1 = p.A

    #Pop 2
    G = alleles[idx.p2]
    no.AA=length(which(G==0))
    no.AB=length(which(G==1))
    no.BB=length(which(G==2))
    n2=(no.AA+no.AB+no.BB)*2
    p.A=(no.AA*2 + no.AB)/n2
    #p.B=(no.BB*2 + no.AB)/n2
    #p2 = min(p.A,p.B,na.rm = T)
    p2 = p.A

    
    N = (p1 - p2)^2 - p1*(1-p1)/(n1-1) - p2*(1-p2)/(n2-1)
    D = p1*(1-p2) + p2*(1-p1)
    
    return(c(N,D))
  }
  
  set.fst = apply(raw.data,2,prestep.fst.one.marker,idx.p1=idx.p1, idx.p2=idx.p2)
  #fst = mean(set.fst[1,],na.rm=T) / mean(set.fst[2,],na.rm=T)
  fst = mean(set.fst[1,]/set.fst[2,],na.rm = T)

  return(fst)
}

#--include--end--#

#24/01/2016
#raw.data is a SNP matrix while rows represent individuals and columns represent SNP
#idx.p1 is a vector containing the indexes of population 1 in raw.data
#idx.p2 is a vector containing the indexes of population 2 in raw.data
#return a vector of Fst
fst.each.snp.hudson <-function(raw.data,idx.p1,idx.p2){
  prestep.fst.one.marker <- function(alleles,idx.p1,idx.p2){
    
    #Pop 1
    G = alleles[idx.p1]
    no.AA=length(which(G==0))
    no.AB=length(which(G==1))
    no.BB=length(which(G==2))
    n1=(no.AA+no.AB+no.BB)*2
    p.A=(no.AA*2 + no.AB)/n1
    #p.B=(no.BB*2 + no.AB)/n1
    #p1 = min(p.A,p.B,na.rm = T)
    p1 = p.A
    
    #Pop 2
    G = alleles[idx.p2]
    no.AA=length(which(G==0))
    no.AB=length(which(G==1))
    no.BB=length(which(G==2))
    n2=(no.AA+no.AB+no.BB)*2
    p.A=(no.AA*2 + no.AB)/n2
    #p.B=(no.BB*2 + no.AB)/n2
    #p2 = min(p.A,p.B,na.rm = T)
    p2 = p.A
    
    
    N = (p1 - p2)^2 - p1*(1-p1)/(n1-1) - p2*(1-p2)/(n2-1)
    D = p1*(1-p2) + p2*(1-p1)
    
    return(c(N,D))
  }
  
  set.fst = apply(raw.data,2,prestep.fst.one.marker,idx.p1=idx.p1, idx.p2=idx.p2)
  #fst = mean(set.fst[1,],na.rm=T) / mean(set.fst[2,],na.rm=T)
  vec.fst = set.fst[1,]/set.fst[2,]
  
  return(vec.fst)
}



R/plinkformat.R
##########################################################################
## IPCAPS Library
## Author: Kridsadakorn Chaichoompu (kridsadakorn.chaichoompu@ulg.ac.be)
## Date: 01/06/2016
## Description:
##    This code is a part of Iterative Pruning to CApture Population 
##    Structure (IPCAPS) Library
##    
##Licence: GPL V3
## 
##    Copyright (C) 2016  Kridsadakorn Chaichoompu
##
##    This program is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    This program is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with this program.  If not, see <http://www.gnu.org/licenses/>.

#--include--begin--#

#For detail about BED file format, check http://pngu.mgh.harvard.edu/~purcell/plink/binary.shtml

# Return object as
# object$snp - SNP matrix from bed file
# object$snp.info - SNP information from bim file
# object$ind.info - individual information from fam file

#Example:
#bed = "test.bed"
#bim = "test.bim"
#fam = "test.fam"
#snp = read.bed(bed,bim,fam)

read.bed <- function(bed,bim,fam,only.snp=FALSE){
  ret = NA
  
  if (!file.exists(bed)){
    cat(paste0("Error: BED file doesn't exist: ",bed,"\n"))
    return(ret)
  }
  if (!file.exists(bim)){
    cat(paste0("Error: BIM file doesn't exist: ",bim,"\n"))
    return(ret)
  }
  if (!file.exists(fam)){
    cat(paste0("Error: FAM file doesn't exist: ",fam,"\n"))
    return(ret)
  }
  #Read BIM file
  snp.info = read.table(bim,header=FALSE)
  colnames(snp.info) = c("chr","ID","GD","position","allele1","allele2")
  
  #Read FAM file
  ind.info = read.table(fam,header=FALSE)
  colnames(ind.info) = c("FamID","IndID","PatID","MatID","sex","phenotype")
  
  no.ind = dim(ind.info)[1]
  no.snp = dim(snp.info)[1]
  
  if (only.snp == TRUE){
    snp.info = NA
    ind.info = NA
  }
  
  #Read BED file
  fh = file(bed, 'rb')
  #Read the first three bytes to check file format
  buff = readBin(fh, what="raw",n=3)
  if (sum(buff[1:2] == c('6c','1b')) != 2){
    cat(paste0("Error: BED file is not in a correct format: ",bed,"\n"))
    return(ret)
  }
  no.byte.to.read = NA
  no.loop = NA
  if (buff[3] == '01'){
    no.byte.to.read = ceiling(no.ind/4.0)
    no.loop = no.snp
  }else{
    no.byte.to.read = ceiling(no.snp/4.0)
    no.loop = no.ind
  }

  snp = readBin(fh, what="raw",n=(no.byte.to.read*no.loop))
  close(fh)
  
  length.snp = length(snp)
  no.part=10
  sub.length.snp = round(length.snp / no.part)
  
  tmp.snp = c()
  for (i in 1:no.part){
    if (i != no.part){
      idx1 = (i-1)*sub.length.snp + 1
      idx2 = i*sub.length.snp
      sub.snp =snp[idx1:idx2]
    }else{
      idx1 = (i-1)*sub.length.snp + 1
      idx2 = length.snp
      sub.snp =snp[idx1:idx2]
    }

    sub.snp.bits = matrix(as.integer(rawToBits(sub.snp)),ncol=2,byrow=T)
    
    tmp.sub.snp.bits = sub.snp.bits[,1]+sub.snp.bits[,2]
    tmp.sub.snp.bits[which((sub.snp.bits[,2]-sub.snp.bits[,1])<0)] = NA
    tmp.snp = c(tmp.snp,tmp.sub.snp.bits)
  }
  
  snp = matrix(tmp.snp,ncol=no.snp,byrow=F)[1:no.ind,]
  
  if (only.snp == FALSE){
    return(list("snp"=snp,"snp.info"=snp.info,"ind.info"=ind.info))
  }else{
    return(list("snp"=snp))
  }
}

# The variable "object" needs to be in this format:
# object$snp - SNP matrix from bed file
# object$snp.info - SNP information from bim file
# object$ind.info - individual information from fam file
# The variable "file" is a prefix for output files
write.bed <- function(object,file){
  
  bed = paste0(file,'.bed')
  bim = paste0(file,'.bim')
  fam = paste0(file,'.fam')
  if (file.exists(bed)){
    cat(paste0("Overwrite the existed bed file\n"))
  }
  if (file.exists(bim)){
    cat(paste0("Overwrite the existed bim file\n"))
  }
  if (file.exists(fam)){
    cat(paste0("Overwrite the existed fam file\n"))
  }
  
  #Write BIM file
  if (is.null(object$snp.info)){
    cat(paste0("Couldn't find object$snp.info: a bim file was not created\n"))
  }else{
    write.table(object$snp.info,file=bim,quote=F,row.names=F,col.names=F)
  }
  
  #Write FAM file
  if (is.null(object$ind.info)){
    cat(paste0("Couldn't find object$ind.info: a fam file was not created\n"))
  }else{
    write.table(object$ind.info,file=fam,quote=F,row.names=F,col.names=F)
  }
  
  #Header of BED file
  header = c(0,0,1,1,0,1,1,0,1,1,0,1,1,0,0,0,1,0,0,0,0,0,0,0)
  
  no.ind = dim(object$ind.info)[1]
  no.snp = dim(object$snp.info)[1]
  
  fill.up = 4-(no.ind %% 4)
  tmp = NA
  if (fill.up != 4){
    tmp = matrix(rep(0,no.snp*fill.up),ncol=no.snp)
  }
  
  vec = object$snp
  if (!anyNA(tmp)){
    vec = rbind(vec,tmp)
  }
  
  vec = as.vector(vec)
  buff = matrix(rep(0,length(vec)*2),ncol=2)
  buff[which(vec == 1),1] = 0
  buff[which(vec == 1),2] = 1
  buff[which(vec == 2),1] = 1
  buff[which(vec == 2),2] = 1
  buff[which(is.na(vec)),1] = 1
  buff[which(is.na(vec)),2] = 0
  vec = NULL
  buff = c(header,as.vector(t(buff)))
  buff=packBits(as.raw(buff))
  
  fh = file(bed, 'wb')
  writeBin(buff,fh)
  close(fh)
}

#--include--end--#




R/FILEST_package.R
#' Fine-Level Structure Simulator
#'
#' A population genetic simulator, which is able to generate synthetic datasets
#' for single-nucleotide polymorphisms (SNP) for multiple populations. The
#' genetic distances among populations can be set according to the Fixation
#' Index (Fst). This tool is able to simulate outlying individuals and missing
#' SNPs can be specified. For Genome-wide association study (GWAS), disease
#' status can be set in desired level according risk ratio.
#'
#' The R package \pkg{FILEST} requires \pkg{KRIS} and \pkg{rARPACK}.
#'
#' Here is the list of functions in the R package \pkg{FILEST}:
#' \itemize{
#' \item \code{\link{cbind_bigmatrix}}
#' \item \code{\link{create.template.setting}}
#' \item \code{\link{demo.filest}}
#' \item \code{\link{filest}}
#' \item \code{\link{rbind_bigmatrix}}
#' }
#'
#' @keywords internal
#' @import rARPACK
#' @import KRIS
"_PACKAGE"
#> [1] "_PACKAGE"

NULL


R/principal.component.R
##########################################################################
## IPCAPS Library
## Author: Kridsadakorn Chaichoompu (kridsadakorn.chaichoompu@ulg.ac.be)
## Date: 01/06/2016
## Description:
##    This code is a part of Iterative Pruning to CApture Population 
##    Structure (IPCAPS) Library
##    
##Licence: GPL V3
## 
##    Copyright (C) 2016  Kridsadakorn Chaichoompu
##
##    This program is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    This program is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with this program.  If not, see <http://www.gnu.org/licenses/>.

#--include--begin--#

###
# Parameters
# X         Data matrix which rows represent samples and columns represent features.
# PCscore  To determine whether scaled principal components will be calculated, 
#           otherwise the eigen vectors will be used instead. Default value is TRUE.
# no.pc     A number  of principal components (PCs) to be calculated. Default value is NA,
#           it means that all PCs will be calculated.
# data.type To determine the type of data. It can be "linear" (default) and "snp".
# weight    To determine whether weights will be used to solve correlation between
#           features.
#
# Examples:
#
# PC = cal.PC.linear(X)
# plot(PC$PC)
# plot(PC$evalue)
#
# PC = cal.PC.linear(X,PCscore = FALSE)
# plot(PC$PC)
# plot(PC$evalue)
#
# PC = cal.PC.linear(X,no.pc=3)
# plot(PC$PC)
# plot(PC$evalue)
#
# PC = cal.PC.linear(X, data.type="snp")
# plot(PC$PC)
# plot(PC$evalue)


cal.PC.linear <- function(X, PCscore = TRUE, no.pc = NA, data.type="linear", XXT=TRUE){
  
  replace.missing <- function(X,missing=NA,rep){
    if (is.na(missing)){
      idx = which(is.na(X))
      X[idx] = rep[idx]
    }else{
      idx = which(X == missing)
      X[idx] = rep[idx]
    }
    return(X)
  }
  

  #Resolve missing value by median
  #cat(paste0("Missing values are set as medians\n"))
  X.median = apply(X,2,median,na.rm=TRUE)
  missing.char=NA
  X = t(apply(X,1,replace.missing,missing=missing.char,rep=X.median))
  
  if (data.type == "snp"){
    XPi = (colSums(X)+1)/(2+(2*dim(X)[1]))
    XSD = (XPi * (1-XPi))^0.5
  }else{
    #Assume linear
    XSD = apply(X,2,sd)
  }
  
  XCM = colMeans(X)
  A = t((t(X)-XCM)/XSD)
  A = A[,colSums(is.na(A))<nrow(A)]
  
  if (is.na(no.pc)){
    no.pc = dim(A)[1]
  }
  
  if (XXT == TRUE){
    AA=A %*% t(A)
    UDV = svds(AA, k=no.pc)
    eigen.value = UDV$d
    PCs = UDV$u
  }else{
    UDV = svds(A, k=no.pc)
    eigen.value = UDV$d
    PCs = UDV$u
  }
  
  
  if (PCscore == TRUE){
    #Calculate the PCscore and the scaled-up PCs
    B.T=diag(1/sqrt(eigen.value)) %*% t(PCs) %*% A
    PCs=A %*% t(B.T)
  }
  
  return(list("PC"=PCs,"evalue"=eigen.value))
}

#--include--end--#


R/plot.3views.R
##########################################################################
## IPCAPS Library
## Author: Kridsadakorn Chaichoompu (kridsadakorn.chaichoompu@ulg.ac.be)
## Date: 01/06/2016
## Description:
##    This code is a part of Iterative Pruning to CApture Population 
##    Structure (IPCAPS) Library
##    
##Licence: GPL V3
## 
##    Copyright (C) 2016  Kridsadakorn Chaichoompu
##
##    This program is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    This program is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with this program.  If not, see <http://www.gnu.org/licenses/>.

#--include--begin--#

#col.pat.table, example
#[,1]   [,2] [,3]    
#[1,] "pop1" "x"  "yellow"
#[2,] "pop2" "u"  "blue" 

plot.3views <- function(X,labels,col.pat.table=NA,only.row=NA){  
  map.color = c("red",rgb(0,68,27,max=255),"blue",rgb(231,41,138,max=255),"darkorange","black")
  map.color = c(map.color,rgb(102,37,6,max=255),rgb(63,0,125,max=255),"green")
  map.color = c(map.color,"cyan",rgb(250,159,181,max=255),"yellow","darkgrey")
  map.color = c(map.color,rgb(116,196,118,max=255))
  
  map.pch = c(1,0,2:18,35:38,60:64,94,126)
  map.pch = c(map.pch,33:34,42,45,47,40,91,123,41,92,93,125)
  map.pch = c(map.pch,49:57,97:107,109:110,112:119,121:122)
  map.pch = c(map.pch,65:78,81:82,84:85,89)

  map.pattern = c()
  for (i in 1:length(map.pch))
    for (j in 1:length(map.color)){
      tmp = c(i,j)
      map.pattern = rbind(map.pattern,tmp)
    }

  if (class(labels) == "data.frame"){
    labels = labels[,1]
    u.label = sort(unique(labels))
  }else{
    u.label = sort(unique(labels))
  }
  
  par(mfrow=c(2,2))
  
  #Top-Left
  par(mar=c(4, 1, 1, 2))
  plot(c(min(X[,1]),max(X[,1])),c(min(X[,2]),max(X[,2])),type="n",xlab="",ylab="",main="",axes=FALSE)
  axis(side=1, labels=TRUE, line=2)
  axis(side=4, labels=TRUE, line=2)
  mtext("PC1", side=1, line=0.5)
  mtext("PC2", side=4, line=0.5)
  set_legend = NULL
  set_pch = NULL
  set_col = NULL
  for (k in 1:length(u.label)){
    if (anyNA(col.pat.table)){
      spch = map.pch[map.pattern[k,1]]
      scolor = map.color[map.pattern[k,2]]
    }else{
      idx = which(col.pat.table[,1] == u.label[k])
      spch = col.pat.table[idx,2]
      scolor = col.pat.table[idx,3]
    }
    
    if (anyNA(only.row)){
      points(X[labels %in% u.label[k],1],X[labels %in% u.label[k],2],col=scolor,pch=spch)
    }else{
      tmp.idx1=which(labels %in% u.label[k])
      tmp.idx2=intersect(tmp.idx1,only.row)
      if (length(tmp.idx2)>0){
        points(X[tmp.idx2,1],X[tmp.idx2,2],col=scolor,pch=spch)
      }
    }
    set_pch = c(set_pch,spch)
    set_col = c(set_col, scolor)
  }
  
  #Top-Right
  par(mar=c(4, 3.5, 1, 1))
  plot(c(min(X[,3]),max(X[,3])),c(min(X[,2]),max(X[,2])),type="n",xlab="",ylab="",main="",axes=FALSE)
  axis(side=1, labels=TRUE, line=2)
  mtext("PC3", side=1, line=0.5)
  for (k in 1:length(u.label)){
    if (anyNA(col.pat.table)){
      spch = map.pch[map.pattern[k,1]]
      scolor = map.color[map.pattern[k,2]]
    }else{
      idx = which(col.pat.table[,1] == u.label[k])
      spch = col.pat.table[idx,2]
      scolor = col.pat.table[idx,3]
    }
    if (anyNA(only.row)){
      points(X[labels %in% u.label[k],3],X[labels %in% u.label[k],2],col=scolor,pch=spch)
    }else{
      tmp.idx1=which(labels %in% u.label[k])
      tmp.idx2=intersect(tmp.idx1,only.row)
      if (length(tmp.idx2)>0){
        points(X[tmp.idx2,3],X[tmp.idx2,2],col=scolor,pch=spch)
      }
    }
  }
  
  #Bottom-Left
  par(mar=c(1, 1, 0.5, 2))
  plot(c(min(X[,1]),max(X[,1])),c(min(X[,3]),max(X[,3])),type="n",xlab="",ylab="",main="",axes=FALSE)
  axis(side=4, labels=TRUE, line=2)
  mtext("PC3", side=4, line=0.5)
  for (k in 1:length(u.label)){
    if (anyNA(col.pat.table)){
      spch = map.pch[map.pattern[k,1]]
      scolor = map.color[map.pattern[k,2]]
    }else{
      idx = which(col.pat.table[,1] == u.label[k])
      spch = col.pat.table[idx,2]
      scolor = col.pat.table[idx,3]
    }
    
    if (anyNA(only.row)){
      points(X[labels %in% u.label[k],1],X[labels %in% u.label[k],3],col=scolor,pch=spch)
    }else{
      tmp.idx1=which(labels %in% u.label[k])
      tmp.idx2=intersect(tmp.idx1,only.row)
      if (length(tmp.idx2)>0){
        points(X[tmp.idx2,1],X[tmp.idx2,3],col=scolor,pch=spch)
      }
    }
  }
  
  #Bottom-Right
  plot(1, type = "n", axes=FALSE, xlab="", ylab="")
  if (length(u.label)>54){
    legend('center', inset=0, legend=u.label, pch=set_pch, col=set_col, ncol=4)
  }else if (length(u.label)>36){
    legend('center', inset=0, legend=u.label, pch=set_pch, col=set_col, ncol=3)
  }else if (length(u.label)>18){
    legend('center', inset=0, legend=u.label, pch=set_pch, col=set_col, ncol=2)
  }else{
    legend('center', inset=0, legend=u.label, pch=set_pch, col=set_col, ncol=1)
  } 
}

#--include--end--#
